packages/react-components/react-tree/src/components/Tree/useNestedControllableCheckedItems.ts

```diff
@@ -0,0 +1,70 @@
+import { useControllableState } from '@fluentui/react-utilities';
+import * as React from 'react';
+import type { TreeCheckedChangeData, TreeProps } from './Tree.types';
+import { ImmutableMap } from '../../utils/ImmutableMap';
+import { createCheckedItems } from '../../utils/createCheckedItems';
+import { TreeItemValue } from '../TreeItem/TreeItem.types';
+import { dataTreeItemValueAttrName } from '../../utils/getTreeItemValueFromElement';
+import { HTMLElementWalker } from '../../hooks/useHTMLElementWalker';
+
+export type ControllableCheckedItemsOptions = {
+  checkedItems?: TreeProps['checkedItems'];
+  defaultCheckedItems?: TreeProps['checkedItems'];
+};
+
+export function useNestedControllableCheckedItems(props: ControllableCheckedItemsOptions) {
+  return useControllableState({
+    initialState: ImmutableMap.empty,
+    state: React.useMemo(() => props.checkedItems && createCheckedItems(props.checkedItems), [props.checkedItems]),
+    defaultState: () => createCheckedItems(props.defaultCheckedItems),
+  });
+}
+
+export function createNextNestedCheckedItems(
+  data: TreeCheckedChangeData,
+  previousCheckedItems: ImmutableMap<TreeItemValue, 'mixed' | boolean>,
+  walker: HTMLElementWalker,
+): ImmutableMap<TreeItemValue, 'mixed' | boolean> {
+  if (data.selectionMode === 'single') {
+    return ImmutableMap.create([[data.value, data.checked]]);
+  }
+  const nextCheckedItems = new Map(previousCheckedItems.set(data.value, data.checked));
+  walker.currentElement = data.target;
+  let previousParent = walker.currentElement;
+  while (walker.parentElement() !== null) {
+    const siblings = getSiblings(walker, previousParent);
+    for (const sibling of siblings) {
+      const value = sibling.getAttribute(dataTreeItemValueAttrName) as TreeItemValue;
+      const ariaChecked = sibling.getAttribute('aria-checked') as 'true' | 'false' | undefined;
+      nextCheckedItems.set(value, !ariaChecked ? 'mixed' : ariaChecked === 'true');
+    }
+    const isAllSiblingsEqualSelectionState = siblings.every(item => {
+      const value = item.getAttribute(dataTreeItemValueAttrName) as TreeItemValue;
+      return nextCheckedItems.get(value) === data.checked;
+    });
+    nextCheckedItems.set(
+      walker.currentElement.getAttribute(dataTreeItemValueAttrName) as TreeItemValue,
+      isAllSiblingsEqualSelectionState ? data.checked : 'mixed',
+    );
+    previousParent = walker.currentElement;
+  }
+  data.target.querySelectorAll('[role="treeitem"]').forEach(item => {
```

Comment:
you already have a walker here, you could just set the currentNode to the target and walk till the end